### ISSUE TITLE
Gradient null for simple calculation

### ISSUE THREAD

**Issue by roelofsmn** [dotnet/TorchSharp#1116]
I wrote an example using TorchSharp to compute the gradient of a simple function as follows:
```csharp
var device = torch.cuda_is_available() ? torch.CUDA : torch.CPU;

torch.Tensor x1 = torch.tensor(new double[] { 1.0, 3.0 }, device: device, requires_grad: true);

torch.Tensor intercepts = torch.tensor(new double[] { 1.0, 5.0 }, device: device, requires_grad: true);

var x2 = x1 * x1;
var example = x2.sum();
example.backward();
var gradients = x1.grad()?.data<double>().ToArray();
Console.WriteLine($"Gradients at minimum: dy/x1={gradients?[0]}, dy/dx2={gradients?[1]}");
```

However, the gradients variable is `null`.

When I create two variables x1 and x2 as scalar tensors, I can get the expected gradients for each.

Does someone know why the gradients in above code are `null`?

P.S. I used TorchSharp 0.100.5 and TorchSharp-cuda-windows 0.100.5

**Comment by roelofsmn** [comment]
Update: when setting `device = torch.CPU`, everything works as expected.

**Comment by NiklasGustafsson** [comment]
So, when it's on CUDA, it fails, but not on CPU? What happens if you move x1 to CPU before calling `grad()`?

To clarify -- it's the tensor coming out of the call to `grad()` that is null, not just the `gradients` array.

**Comment by roelofsmn** [comment]
Correct. If I do
```
var example = objectiveFunction(x1, intercepts);
example.backward();
x1 = x1.to(torch.CPU);
var gradients = x1.grad()?.data<double>().ToArray();
```
also evaluates `gradients` to null. It is indeed the tensor coming out of `grad()` that is null.

**Comment by roelofsmn** [comment]
I'm now getting the following warning:
```
Warning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (function grad)
```
But what makes a tensor a Leaf tensor? Or, rather, why would `x1` not be a leaf tensor in my example?

**Comment by NiklasGustafsson** [comment]
That doesn't make any sense. 

`x1` is a leaf tensor (not computed from something else), while `example` and `x2` aren't. Maybe the call to `to()` turns it into one...

**Comment by roelofsmn** [comment]
Yeah when I added `retain_grad()` the call to `to()` created a new Tensor that is not a leaf and has no gradients. But the original tensor did.
I added the dimension along which to sum (`(long)0`), but that does not help, nor does adding the dtype explicitly.

**Comment by NiklasGustafsson** [comment]
@roelofsmn -- still tracking down everywhere this bug is manifested, but for now, try this alternative to creating the parameter:

```C#
var x1 = torch.tensor(new double[] { 1.0, 3.0 }, device: device).with_requires_grad(true);
```
